omod/src/test/java/org/openmrs/web/controller/ForgotPasswordFormControllerTest.java

```diff
@@ -0,0 +1,281 @@
+		
+		
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		MockHttpServletResponse response = new MockHttpServletResponse();
+		
+		
+		request.setParameter("uname", "bruno");
+		request.setParameter("secretAnswer", "valid secret answer");
+		
+		request.setParameter("uname", "validuser");
+		
+		request.setMethod("POST");
+		
+		controller.handleRequest(request, response);
+		
+		
+		Assert.assertEquals("validuser", request.getAttribute("uname"));
+		Assert.assertEquals("valid secret question", request.getAttribute("secretQuestion"));
+	}
+	
+	
+	@Test
+	public void shouldAuthenticateAsUserWithValidSecretQuestion() throws Exception {
+		ForgotPasswordFormController controller = (ForgotPasswordFormController) applicationContext.getBean("forgotPasswordForm");
+		
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		
+		request.addParameter("uname", "validuser");
+		request.addParameter("secretAnswer", "valid secret answer");
+		
+		HttpServletResponse response = new MockHttpServletResponse();
+		controller.handleRequest(request, response);
+		
+		Assert.assertEquals(2, Context.getAuthenticatedUser().getId().intValue());
+	}
+	
+	/**
+	 * If a user enters the wrong secret answer, they should be kicked back to the form and not be
+	 * authenticated
+	 *
+	 * @throws Exception
+	 */
+	@Test
+	public void shouldNotAuthenticateAsUserWithValidSecretQuestion() throws Exception {
+		ForgotPasswordFormController controller = (ForgotPasswordFormController) applicationContext.getBean("forgotPasswordForm");
+		
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		
+		request.addParameter("uname", "validuser");
+		request.addParameter("secretAnswer", "invalid secret answer");
+		
+		HttpServletResponse response = new MockHttpServletResponse();
+		controller.handleRequest(request, response);
+		
+		Assert.assertFalse(Context.isAuthenticated());
+	}
+	
+	/**
+	 * If a user enters 5 requests, the 6th should fail even if that one has a valid username in it
+	 *
+	 * @throws Exception
+	 */
+	@Test
+	public void shouldLockOutAfterFiveFailedInvalidUsernames() throws Exception {
+		ForgotPasswordFormController controller = (ForgotPasswordFormController) applicationContext.getBean("forgotPasswordForm");
+		
+		for (int x = 1; x <= 5; x++) {
+			MockHttpServletRequest request = new MockHttpServletRequest();
+			request.setMethod("POST");
+			
+			request.addParameter("uname", "invaliduser");
+			
+			controller.handleRequest(request, new MockHttpServletResponse());
+		}
+		
+		// those were the first five, now the sixth request (with a valid user) should fail
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.setMethod("POST");
+		
+		request.addParameter("uname", "validuser");
+		
+		controller.handleRequest(request, new MockHttpServletResponse());
+		Assert.assertNull(request.getAttribute("secretQuestion"));
```

Comment:
You can have a secret question when you are not locked out, if you're saying the user is locked in the test method name out then you need to check that they are actually locked out in the database, it's importing for test method name to be reflective of the actual thing you are testing